daft/table/table.py

```diff
@@ -71,9 +71,36 @@ def _from_pytable(pyt: _PyTable) -> Table:
     @staticmethod
     def from_arrow(arrow_table: pa.Table) -> Table:
         assert isinstance(arrow_table, pa.Table)
+        arrow_schema = arrow_table.schema
         schema = Schema._from_field_name_and_types(
             [(f.name, DataType.from_arrow_type(f.type)) for f in arrow_table.schema]
         )
+        # TODO(Clark): For pyarrow < 12.0.0, struct array slice offsets are dropped
+        # when converting to record batches. We work around this below by flattening
+        # the field arrays for all struct arrays, which propagates said offsets to
+        # the child (field) arrays. Note that this flattening is zero-copy unless
+        # the column contains a validity bitmap (i.e. the column has one or more nulls).
+        #
+        # We should enforce a lower Arrow version bound once the upstream fix is
+        # released.
+        #
+        # See:
+        #  - https://github.com/apache/arrow/issues/34639
+        #  - https://github.com/apache/arrow/pull/34691
+        for idx, name in enumerate(arrow_schema.names):
+            field = arrow_schema.field(name)
+            dtype = field.type
+            if pa.types.is_struct(dtype):
+                fields = [dtype.field(i) for i in range(dtype.num_fields)]
+                new_chunks = []
+                # Flatten each chunk to propagate slice offsets to child arrays.
+                for chunk in arrow_table.column(name).chunks:
+                    new_field_arrays = chunk.flatten()
+                    new_chunk = pa.StructArray.from_arrays(new_field_arrays, fields=fields)
+                    new_chunks.append(new_chunk)
+                new_column = pa.chunked_array(new_chunks)
+                arrow_table = arrow_table.set_column(idx, field, new_column)
```

Comment:
@samster25 @jaychia Here is the workaround for `StructArray` slice offsets being dropped when converting to record batches. This is essentially the yet-to-be-released Arrow Core fix in the linked PR.

Reply:
Nice!

Reply:
1. Could we pull this out into a separate helper?
2. The helper could no-op `if pa.__version__ >= 12.0`
3. Is this behavior testable? We should add a unit test so that when pyarrow does upgrade to 12.0 we can verify that a no-op actually works with that version

Lmk what you think!

Reply:
1. Sure!
2. Assuming we're going to want to support pyarrow versions older than 12.0.0 at that point, that sounds good!
3. Yep, it's currently covered by our [`test_parquet_reads_limit_rows` test in `test_table_io.py`](https://github.com/Eventual-Inc/Daft/blob/c41c574b7a5d065371037fd28fa286bdacf01285/tests/table/test_table_io.py#L156-L160), since our I/O layer slices the pyarrow data before handing it off to our Arrow2 backend. I'll add a quick dedicated test for this issue in `test_from_py.py` as well.